### ISSUE TITLE
[MusicBERT] Restriction to 1002 octuples when using `preprocess.encoding_to_str`

### ISSUE THREAD

**Issue by tripathiarpan20** [microsoft/muzic#60]
Hi once again!

While preprocessing a MIDI file, I noticed that the `MIDI_to_encoding` method performs as intended and converts the sample song to 106 bars as seen in the snip below of the resultant octuples (please correct me if I'm wrong).

However, the `encoding_to_str` method has the result with restriction to just 18 bars (as conculsive from highlighted `<0-18>` near the end of the encoded string in snip below):

![image](https://user-images.githubusercontent.com/42506819/177249705-ceedd6b1-b716-4959-92b4-f89ed4863ff2.png)

More generally, what I have noticed in cases of multiple MIDI files is that only upto the first 1000 octuples (*i.e, start token octuple + 1000 note octuples + end token octuple = (1002 * 8) = 8016 tokens*) are considered:

![image](https://user-images.githubusercontent.com/42506819/177250055-9cac1403-3a3d-4033-96af-dc8129f98d34.png)

Is there any way to change `encoding_to_str` to get the whole song instead?, upto 256 bars only I mean, as model vocabulary is also restricted to 256 bars. 
I am not familiar enough with miditoolkit or mido to understand the code properly as of now, else I would have tried to fix this.

Thanks in advance! 

Edit: I am aware that the `musicbert_base` model can support upto 8192 octuples (i.e, final input to MusicBERT encoder) only, but that does not seem to be the issue here I think.

**Comment by mlzeng** [comment]
`musicbert_base` model can only support up to 1024 octuples (= 8192 tokens)

If you need to change `encoding_to_str` to get the whole song, you can increase the value of `sample_len_max` or just replace the `for i in e[p: p + sample_len_max]` code in that function with `for i in e`.

**Comment by tripathiarpan20** [comment]
Hi @mlzeng .
There is a bit of confusion, with context from [`musicbert/musicbert/__init__.py`](https://github.com/microsoft/muzic/blob/main/musicbert/musicbert/__init__.py), let's say we have the input `tokens` of some batch size having 8192 octuples (i.e, 8192 * 8 tokens) in each sample. 

`tokens` would be something like `torch.Tensor([ [0, 0, 0, 0, 0, 0, 0, 0, ............., 2, 2, 2, 2, 2, ,2, 2, 2].... ])` given that token mapping of `<s>` is `0` and `</s>` is `2` obtained from `roberta_base.task.label_dictionary.index('<s>')` and so on. 
Also, `tokens.shape[1] % 8 == 0` is True because Octuples are there. 

The `OctupleEncoder` in [lines 251-252](https://github.com/microsoft/muzic/blob/main/musicbert/musicbert/__init__.py#L251-L252) first convert `tokens` of dimension `[batch_size, 8192 * 8]` to `[batch_size, 8192 * 8, 768]` as `x` (see green arrow in Colab snip at end of comment) :

![image](https://user-images.githubusercontent.com/42506819/177278659-b86bb53b-f81c-4dc2-a9ac-413c0c539b99.png)


The [lines 253-254](https://github.com/microsoft/muzic/blob/main/musicbert/musicbert/__init__.py#L253-L254) then convert `x` from dimension `[batch_size, 8192 * 8, 768]` to dimension  `[batch_size, 8192, 768]`  (`ratio` is 8 from Line 232) (see red arrow in Colab snip at end of comment):
![image](https://user-images.githubusercontent.com/42506819/177280010-acb13232-b40e-4293-89a8-0e034d6e7457.png)

The [lines 257-259](https://github.com/microsoft/muzic/blob/main/musicbert/musicbert/__init__.py#L257-L259) then add positional embedding using only every (8*i + 1) th element of `x` (since ratio is 8), which is the `measure` (i.e, bar) field of **each** octuple:
![image](https://user-images.githubusercontent.com/42506819/177282162-f9ccfff3-95c9-456c-8ae3-6f0ef398dc61.png)
![image](https://user-images.githubusercontent.com/42506819/177280876-973ac95b-d4a2-40b1-8f26-a365937502db.png) 

Now if we see the definition of `embed_positions` in MusicBERTEncoder marked by blue arrow in Colab snip below, `tokens[:, ::ratio].shape[1]` should be equal to `8194`!!
 
![image](https://user-images.githubusercontent.com/42506819/177279021-f01661e4-cea1-4051-845b-6b58c187aaf5.png)
![image](https://user-images.githubusercontent.com/42506819/177280149-6da8bbe7-acd0-4f96-b6da-f53a062bc379.png)

Therefore upto 8194 octuples should be supported (could be with out without padding)!

Please let me know if there is any misunderstanding in my logic, as I really need to understand this.

Thanks!

**Comment by mlzeng** [comment]
Hi @tripathiarpan20 

Your logic is pretty correct. But MusicBERT is trained with setting `TOKENS_PER_SAMPLE=8192` (as seen in `train_mask.sh`) which means the length of input sequences would not exceed 8192 tokens (= 1024 octuples), and the attention layers in the encoder will only get tensors with length no more than 1024.

Processing 8192 octuple tokens with MusicBERT is theoretically possible, but that will require 64 times more GPU memory, which is non-practical currently.

**Comment by tripathiarpan20** [comment]
I see, it is now clear! :smiley:

The provided MusicBERT Base also performs amazing without any further training and thanks to the team once again for making it open-source once.

In `musicbert/musicbert/__init__.py` I also noticed definition of `musicbert_large`. Is that another model that will be released eventually?

**Comment by tripathiarpan20** [comment]
Oh I see, hoping it is resolved eventually.

Would the finetuned models on the genre prediction and accompaniment suggestion be released too? 
I'm thinking of implementing a genre based task and the finetuned model would be a great starting point.
